src/convnets/resnets/core.jl

```diff
@@ -0,0 +1,337 @@
+        elseif stem_type == :deep_tiered
+            stem_channels = (3 * (stem_width ÷ 4), stem_width)
+        end
+        conv1 = Chain(conv_norm((3, 3), inchannels => stem_channels[1], activation;
+                                norm_layer, prenorm, stride = 2, pad = 1, bias = false)...,
+                      conv_norm((3, 3), stem_channels[1] => stem_channels[2], activation;
+                                norm_layer, pad = 1, bias = false)...,
+                      Conv((3, 3), stem_channels[2] => inplanes; pad = 1, bias = false))
+    else
+        conv1 = Conv((7, 7), inchannels => inplanes; stride = 2, pad = 3, bias = false)
+    end
+    bn1 = norm_layer(inplanes, activation)
+    # Stem pooling
+    stempool = replace_pool ?
+               Chain(conv_norm((3, 3), inplanes => inplanes, activation; norm_layer,
+                               prenorm,
+                               stride = 2, pad = 1, bias = false)...) :
+               MaxPool((3, 3); stride = 2, pad = 1)
+    return Chain(conv1, bn1, stempool)
+end
+
+resnet_planes(stage_idx::Integer) = 64 * 2^(stage_idx - 1)
+
+function basicblock_builder(block_repeats::Vector{<:Integer}; inplanes::Integer = 64,
+                            reduction_factor::Integer = 1, expansion::Integer = 1,
+                            norm_layer = BatchNorm, prenorm::Bool = false,
+                            activation = relu, attn_fn = planes -> identity,
+                            drop_block_rate = 0.0, drop_path_rate = 0.0,
+                            stride_fn = get_stride, planes_fn = resnet_planes,
+                            downsample_tuple = (downsample_conv, downsample_identity))
+    pathschedule = linear_scheduler(drop_path_rate; depth = sum(block_repeats))
+    blockschedule = linear_scheduler(drop_block_rate; depth = sum(block_repeats))
+    # closure over `idxs`
+    function get_layers(stage_idx::Integer, block_idx::Integer)
+        planes = planes_fn(stage_idx)
+        # `get_stride` is a callback that the user can tweak to change the stride of the
+        # blocks. It defaults to the standard behaviour as in the paper
+        stride = stride_fn(stage_idx, block_idx)
+        downsample_fn = (stride != 1 || inplanes != planes * expansion) ?
+                        downsample_tuple[1] : downsample_tuple[2]
+        # DropBlock, DropPath both take in rates based on a linear scaling schedule
+        schedule_idx = sum(block_repeats[1:(stage_idx - 1)]) + block_idx
+        drop_path = DropPath(pathschedule[schedule_idx])
+        drop_block = DropBlock(blockschedule[schedule_idx])
+        block = basicblock(inplanes, planes; stride, reduction_factor, activation,
+                           norm_layer, prenorm, attn_fn, drop_path, drop_block)
+        downsample = downsample_fn(inplanes, planes * expansion; stride)
+        # inplanes increases by expansion after each block
+        inplanes = planes * expansion
```

Comment:
We need this, though. This is calculating the change in `inplanes` across blocks

Reply:
Maybe I am missing something but I don't see where the output of this calculation goes? It seems unused...unless it is modifying a global which is very bad.

Reply:
We were before, unfortunately. I've pushed a change. This makes `resnet_planes` return a vector instead of being a `stage_idx` based callback - the reason we need this is because `inplanes` needs the `planes` from the _previous_ block, not the current one, so we need to have access to that information